<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>spatial_analysis.r – Spatial transcriptomics analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-a7b8f1f616b980bd82773e6305085fe4.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Spatial transcriptomics analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./notebooks/01_spatial_analysis.R.ipynb"> 
<span class="menu-text">Spatial preprocessing</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#sample-preprocessing" id="toc-sample-preprocessing" class="nav-link active" data-scroll-target="#sample-preprocessing">Sample preprocessing</a></li>
  <li><a href="#preprocess-all-samples" id="toc-preprocess-all-samples" class="nav-link" data-scroll-target="#preprocess-all-samples">Preprocess all samples</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"><h1 class="title display-7"></h1></header><div class="quarto-title-block"><div class="quarto-title-tools-only"><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li></ul></div></div>




<p>Microglial INPP5D limits plaque formation and glial reactivity in the PSAPP mouse model of Alzheimer’s disease</p>
<ul>
<li>Paper: DOI: 10.1002/alz.12821</li>
<li>Aim of the analysis:
<ul>
<li>Annotate the data</li>
<li>Pathway analysis</li>
</ul></li>
</ul>
<div id="cell-2" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="1">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"SpatialExperiment"</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"SingleCellExperiment"</span>))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"DropletUtils"</span>))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"Seurat"</span>))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"ggspavis"</span>))</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"scater"</span>))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"ggplot2"</span>))</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"patchwork"</span>))</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"biomaRt"</span>))</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"nnSVG"</span>))</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"scran"</span>))</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"igraph"</span>))</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"glue"</span>))</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># suppressMessages(library("BayesSpace"))</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># suppressMessages(library("pheatmap"))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“package ‘DropletUtils’ was built under R version 4.4.3”
Warning message:
“package ‘SeuratObject’ was built under R version 4.4.3”</code></pre>
</div>
</div>
<div id="cell-3" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # another options for vscode for plotting</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># options(repr.plot.width = 10, repr.plot.height = 5, jupyter.plot_scale = 1)</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># # make the resolution better</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.res =</span> <span class="dv">300</span>) <span class="co"># set dpy to 200</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-4" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># options(repr.plot.width = , repr.plot.height = "auto", jupyter.plot_scale = 1, repr.plot.res = "auto")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-5" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>main_folder <span class="ot">&lt;-</span> <span class="st">"/home/kharuk/projects/courses/st_daad/data"</span>  <span class="co"># Change to your actual path</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># List all sample directories inside the main folder</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>sample_dirs <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(main_folder, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(sample_dirs)  <span class="co"># Check detected directories</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-6" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="st">"GSM6171784_PSAPP_CO1"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-7" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upload one sample for preprocessing</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">read10xVisium</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">samples =</span> sample_dirs[<span class="dv">1</span>],</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">sample_id =</span> <span class="st">"GSM6171784_PSAPP_CO1"</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">data =</span> <span class="st">"filtered"</span>,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">type =</span> <span class="st">"sparse"</span>,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">images =</span> <span class="st">"lowres"</span>,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">load =</span> <span class="cn">TRUE</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-8" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move rownames to rowData names new column ensembl_id</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>ensembl_id <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">rowData</span>(spe))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="sample-preprocessing" class="level3">
<h3 class="anchored" data-anchor-id="sample-preprocessing">Sample preprocessing</h3>
<section id="plot-data" class="level4">
<h4 class="anchored" data-anchor-id="plot-data">Plot data</h4>
<div id="cell-11" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot spatial coordinates (spots)</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpots</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="qc-metrics" class="level4">
<h4 class="anchored" data-anchor-id="qc-metrics">QC metrics</h4>
<div id="cell-13" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First, we subset the object to keep only spots over tissue.</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the object to keep only spots over tissue</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="fu">colData</span>(spe)<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-14" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-15" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># identify mito genes</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"(^MT-|^mt-)"</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>symbol)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(is_mito)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-16" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># print mito genes</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>symbol[is_mito]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-17" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate per-spot QC metrics and store in coldata </span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(spe, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(spe))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="selecting-thresholds-for-filtering" class="level4">
<h4 class="anchored" data-anchor-id="selecting-thresholds-for-filtering">Selecting thresholds for filtering</h4>
<p>Have to select thresholds for several QC metrics: (i) library size, (ii) number of expressed genes, (iii) proportion of mitochondrial reads, and (iv) number of cells per spot (does not exist in data).</p>
<p>Library size represents the total sum of UMI counts per spot.</p>
<div id="cell-21" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>sum, <span class="at">breaks =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-22" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-23" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># histograms of QC metrics before filtering</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co"># hist(colData(spe)$cell_count, xlab = "number of cells", main = "No. cells per spot")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-24" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># asign gene names for plotting </span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(spe) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>symbol</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(spe))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-25" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the same QC on the spatial coordinates</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>pal <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#fde725"</span>, <span class="st">"navy"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"sum"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span>  </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"detected"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()a</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>qc_before <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(p1, p2, p3, <span class="at">nrow =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"QC metrics for {sample_id} before filtering"</span>))</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>qc_before</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">glue</span>(<span class="st">"qc_metrics_{sample_id}_before_filtering.png"</span>), <span class="at">plot =</span> qc_before, <span class="at">width =</span> <span class="dv">10</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-26" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot QC with image included</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">annotate =</span> <span class="st">"sum"</span>, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">point_size =</span> <span class="fl">0.75</span>) <span class="sc">+</span> </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">annotate =</span> <span class="st">"detected"</span>, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">point_size =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">annotate =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">point_size =</span> <span class="fl">0.75</span>) <span class="sc">+</span> </span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>tissue_plot <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">spots =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="co"># display panels using patchwork</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>final_plot <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(p1, p2, p3, tissue_plot, <span class="at">nrow =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"QC metrics for {sample_id} before filtering"</span>))</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">glue</span>(<span class="st">"qc_metrics_{sample_id}_tissue.png"</span>), final_plot, <span class="at">width =</span> <span class="dv">20</span>, <span class="at">height =</span> <span class="dv">5</span>, <span class="at">dpi =</span> <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Another option is to calculate MAD and filter based on this value.</p>
<div id="cell-28" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># determine outliers via thresholding on MAD from the median</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>ol <span class="ot">&lt;-</span> <span class="fu">perCellQCFilters</span>(spe, <span class="at">sub.fields=</span><span class="st">"subsets_mito_percent"</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># add results as cell metadata</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(spe)[<span class="fu">names</span>(ol)] <span class="ot">&lt;-</span> ol </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># tabulate # and % of cells that'd </span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co"># be discarded for different reasons</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">check.names=</span><span class="cn">FALSE</span>,</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">#</span><span class="st">`</span><span class="ot">=</span><span class="fu">apply</span>(ol, <span class="dv">2</span>, sum), </span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">%</span><span class="st">`</span><span class="ot">=</span><span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span><span class="fu">apply</span>(ol, <span class="dv">2</span>, mean), <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-29" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot what spots will be discarded with MAD</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">names</span>(ol), \(.) </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotSpots</span>(spe, <span class="at">annotate=</span>.) <span class="sc">+</span> <span class="fu">ggtitle</span>(.)) <span class="sc">|&gt;</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">wrap_plots</span>(<span class="at">nrow=</span><span class="dv">1</span>, <span class="at">guides=</span><span class="st">"collect"</span>) <span class="sc">+</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"Spots flagged for discard {sample_id}"</span>)) <span class="sc">&amp;</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">col=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">3</span>))) <span class="sc">&amp;</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="st">"discard"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="st">"lavender"</span>, <span class="st">"purple"</span>)) <span class="sc">&amp;</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="dv">8</span>), </span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.key.size=</span><span class="fu">unit</span>(<span class="fl">0.8</span>, <span class="st">"lines"</span>)</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>I’d like not to remove the outliers with MAD, but to set a threshold for each of the QC metrics: library size (sum), number of expressed genes (detected), proportion of mitochondrial reads (subset_mito_percent)</p>
<div id="cell-31" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the threshold manually </span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># select QC thresholds</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">600</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">400</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co"># qc_mito &lt;- colData(spe)$subsets_mito_percent &gt; 30</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co"># number of discarded spots for each metric</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(<span class="fu">cbind</span>(qc_lib_size, qc_detected), <span class="dv">2</span>, sum)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-32" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># combined set of discarded spots</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(discard)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-33" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># store in object</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Plot the set of discarded spots in the spatial x-y coordinates, to confirm that the spatial distribution of the discarded spots does not correspond to any biologically meaningful regions, which would indicate that we are removing biologically informative spots.</p>
<div id="cell-35" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpotQC</span>(spe, <span class="at">plot_type =</span> <span class="st">"spot"</span>, <span class="at">annotate =</span> <span class="st">"discard"</span>) <span class="sc">+</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Spots to be discarded"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-36" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter low-quality spots</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="sc">!</span><span class="fu">colData</span>(spe)<span class="sc">$</span>discard]</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-37" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the same QC on the spatial coordinates</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"sum"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span>  </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"detected"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_void</span>()</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>qc_after <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(p1, p2, p3, <span class="at">nrow =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"QC metrics for {sample_id} before filtering"</span>))</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>qc_after</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">glue</span>(<span class="st">"qc_metrics_{sample_id}_after_filtering.png"</span>), <span class="at">plot =</span> qc_after, <span class="at">width =</span> <span class="dv">10</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-38" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># p1 &lt;- plotVisium(spe, annotate = "sum", highlight = "in_tissue", image = FALSE, point_size = 1)</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co"># p2 &lt;- plotVisium(spe, annotate = "subsets_mito_percent", highlight = "in_tissue", image = FALSE, point_size = 1)</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co"># p3 &lt;- plotVisium(spe, annotate = "Cst7", highlight = "in_tissue", image = FALSE, point_size = 1)</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="co"># wrap_plots(p1, p2, p3, ncol = 2) + </span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="co">#     plot_annotation(title = "QC metrics after filtering")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="normalization-feature-selection-dr-clustering-and-visualization" class="level4">
<h4 class="anchored" data-anchor-id="normalization-feature-selection-dr-clustering-and-visualization">Normalization, Feature selection, DR, Clustering and Visualization</h4>
<section id="non-spatial-normalization" class="level5">
<h5 class="anchored" data-anchor-id="non-spatial-normalization">Non-spatial Normalization</h5>
<div id="cell-41" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate library size factors</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">computeLibraryFactors</span>(spe)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">sizeFactors</span>(spe))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-42" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">sizeFactors</span>(spe), <span class="at">breaks =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-43" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate logcounts and store in object</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">assayNames</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="feature-selection" class="level4">
<h4 class="anchored" data-anchor-id="feature-selection">Feature selection</h4>
<section id="non-spatial-feature-selection" class="level5">
<h5 class="anchored" data-anchor-id="non-spatial-feature-selection">Non-spatial Feature selection</h5>
<div id="cell-46" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove mitochondrial genes</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">!</span>is_mito, ]</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-47" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fit mean-variance relationship</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(spe)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize mean-variance relationship</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">metadata</span>(dec)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fit<span class="sc">$</span>mean, fit<span class="sc">$</span>var, </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"mean of log-expression"</span>, <span class="at">ylab =</span> <span class="st">"variance of log-expression"</span>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="fu">curve</span>(fit<span class="sc">$</span><span class="fu">trend</span>(x), <span class="at">col =</span> <span class="st">"dodgerblue"</span>, <span class="at">add =</span> <span class="cn">TRUE</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-48" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select top HVGs</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(top_hvgs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="dimensionality-reduction" class="level4">
<h4 class="anchored" data-anchor-id="dimensionality-reduction">Dimensionality reduction</h4>
<div id="cell-50" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run PCA and UMAP</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(spe, <span class="at">subset_row=</span>top_hvgs)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred=</span><span class="st">"PCA"</span>, <span class="at">pca =</span> <span class="dv">30</span>) <span class="co"># use 30 PCs as in the paper</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="clustering" class="level5">
<h5 class="anchored" data-anchor-id="clustering">Clustering</h5>
<div id="cell-52" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># build shared nearest-neighbor (SNN) graph</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(spe, <span class="at">use.dimred=</span><span class="st">"PCA"</span>, <span class="at">type=</span><span class="st">"jaccard"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="co"># cluster via Leiden community detection algorithm</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">cluster_leiden</span>(g, <span class="at">objective_function=</span><span class="st">"modularity"</span>, <span class="at">resolution=</span><span class="fl">0.8</span>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(spe<span class="sc">$</span>Leiden <span class="ot">&lt;-</span> <span class="fu">factor</span>(k<span class="sc">$</span>membership))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-53" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>spe</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-54" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot clusters in spatial x-y coordinates</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>spatial_p1 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"Leiden"</span>, <span class="at">legend_position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">unname</span>(pals<span class="sc">::</span><span class="fu">trubetskoy</span>()))  <span class="co"># Custom color palette for spatial plot</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>umap_p2 <span class="ot">&lt;-</span> <span class="fu">plotDimRed</span>(spe, <span class="at">plot_type =</span> <span class="st">"UMAP"</span>, </span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">annotate =</span> <span class="st">"Leiden"</span>) <span class="sc">+</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">unname</span>(pals<span class="sc">::</span><span class="fu">trubetskoy</span>()))  <span class="co"># Custom color palette for spatial plot</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(spatial_p1, umap_p2, <span class="at">nrow=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"Leiden clusters for {sample_id}"</span>))</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-55" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save the object</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(spe, <span class="at">file =</span> <span class="fu">glue</span>(<span class="st">"../data/ST_{sample_id}.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="spatially-aware-spatially-variable-genes-svgs.-eternally-running" class="level5">
<h5 class="anchored" data-anchor-id="spatially-aware-spatially-variable-genes-svgs.-eternally-running">Spatially-aware: Spatially variable genes (SVGs). <em>Eternally running :)</em></h5>
<div id="cell-58" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # filter low-expressed and mitochondrial genes</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># # using very stringent filtering parameters for faster runtime in this example</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co"># # note: for a full analysis, use alternative filtering parameters (e.g. defaults)</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co"># # re-calculate logcounts after filtering</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co"># # using library size factors</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>spe_nnSVG <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co"># # run nnSVG</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co"># # using a single core for compatibility on build system</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="co"># # note: for a full analysis, use multiple cores</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>spe_nnSVG <span class="ot">&lt;-</span> <span class="fu">nnSVG</span>(spe_nnSVG, <span class="at">n_threads =</span> <span class="dv">6</span>)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a><span class="co"># # investigate results</span></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a><span class="co"># # show results</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(spe_nnSVG), <span class="dv">3</span>)</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a><span class="co"># # number of significant SVGs</span></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>padj <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a><span class="co"># # show results for top n SVGs</span></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(spe_nnSVG)[<span class="fu">order</span>(<span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>rank)[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], ]</span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a><span class="co"># # identify top-ranked SVG</span></span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>symbol[<span class="fu">which</span>(<span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>rank <span class="sc">==</span> <span class="dv">1</span>)]</span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
</section>
<section id="preprocess-all-samples" class="level2">
<h2 class="anchored" data-anchor-id="preprocess-all-samples">Preprocess all samples</h2>
<div id="cell-61" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="2">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upload samples</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>main_folder <span class="ot">&lt;-</span> <span class="st">"/home/kharuk/projects/courses/st_daad/data"</span>  <span class="co"># Change to your actual path</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="co"># List all sample directories inside the main folder</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>sample_dirs <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(main_folder, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(sample_dirs)  <span class="co"># Check detected directories</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/home/kharuk/projects/courses/st_daad/data/GSM6171784_PSAPP_CO1" 
[2] "/home/kharuk/projects/courses/st_daad/data/GSM6171785_PSAPP_TAM1"
[3] "/home/kharuk/projects/courses/st_daad/data/GSM6171788_PSAPP_CO2" 
[4] "/home/kharuk/projects/courses/st_daad/data/GSM6171789_PSAPP_TAM2"
[5] "/home/kharuk/projects/courses/st_daad/data/GSM6171792_PSAPP_CO3" 
[6] "/home/kharuk/projects/courses/st_daad/data/GSM6171793_PSAPP_TAM3"</code></pre>
</div>
</div>
<div id="cell-62" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="9">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># function to preprocess a single sample</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>preprocess_sample <span class="ot">&lt;-</span> <span class="cf">function</span>(sample_path) {</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    sample_id <span class="ot">&lt;-</span> <span class="fu">basename</span>(sample_path)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> <span class="fu">read10xVisium</span>(</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">samples =</span> sample_path,</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">sample_id =</span> sample_id,</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">data =</span> <span class="st">"filtered"</span>,</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">type =</span> <span class="st">"sparse"</span>,</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">images =</span> <span class="st">"lowres"</span>,</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">load =</span> <span class="cn">TRUE</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># First, we subset the object to keep only spots over tissue.</span></span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># subset the object to keep only spots over tissue</span></span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> spe[, <span class="fu">colData</span>(spe)<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dim</span>(spe)</span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># identify mito genes</span></span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a>    is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"(^MT-|^mt-)"</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>symbol)</span>
<span id="cb44-22"><a href="#cb44-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">table</span>(is_mito)</span>
<span id="cb44-23"><a href="#cb44-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-24"><a href="#cb44-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># calculate per-spot QC metrics and store in coldata </span></span>
<span id="cb44-25"><a href="#cb44-25" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(spe, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb44-26"><a href="#cb44-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">head</span>(<span class="fu">colData</span>(spe))</span>
<span id="cb44-27"><a href="#cb44-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-28"><a href="#cb44-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-29"><a href="#cb44-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># histograms of QC metrics before filtering # </span><span class="al">TODO</span><span class="co"> add title here </span></span>
<span id="cb44-30"><a href="#cb44-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">png</span>(<span class="fu">glue</span>(<span class="st">"qc_hist_{sample_id}.png"</span>), <span class="at">width =</span> <span class="dv">1000</span>, <span class="at">height =</span> <span class="dv">800</span>)</span>
<span id="cb44-31"><a href="#cb44-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>))  <span class="co"># layout</span></span>
<span id="cb44-32"><a href="#cb44-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-33"><a href="#cb44-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb44-34"><a href="#cb44-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb44-35"><a href="#cb44-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb44-36"><a href="#cb44-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-37"><a href="#cb44-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">"QC histograms for {sample_id}"</span>), <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="dv">1</span>)</span>
<span id="cb44-38"><a href="#cb44-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-39"><a href="#cb44-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dev.off</span>()</span>
<span id="cb44-40"><a href="#cb44-40" aria-hidden="true" tabindex="-1"></a>    <span class="co"># asign gene names for plotting </span></span>
<span id="cb44-41"><a href="#cb44-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(spe) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>symbol</span>
<span id="cb44-42"><a href="#cb44-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(spe))</span>
<span id="cb44-43"><a href="#cb44-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-44"><a href="#cb44-44" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot the same QC on the spatial coordinates</span></span>
<span id="cb44-45"><a href="#cb44-45" aria-hidden="true" tabindex="-1"></a>    pal <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#fde725"</span>, <span class="st">"navy"</span>)</span>
<span id="cb44-46"><a href="#cb44-46" aria-hidden="true" tabindex="-1"></a>    p1 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"sum"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span>  </span>
<span id="cb44-47"><a href="#cb44-47" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-48"><a href="#cb44-48" aria-hidden="true" tabindex="-1"></a>    p2 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"detected"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb44-49"><a href="#cb44-49" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-50"><a href="#cb44-50" aria-hidden="true" tabindex="-1"></a>    p3 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb44-51"><a href="#cb44-51" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-52"><a href="#cb44-52" aria-hidden="true" tabindex="-1"></a>    qc_before <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(p1, p2, p3, <span class="at">nrow =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb44-53"><a href="#cb44-53" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"QC metrics for {sample_id} before filtering"</span>))</span>
<span id="cb44-54"><a href="#cb44-54" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(qc_before)</span>
<span id="cb44-55"><a href="#cb44-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">glue</span>(<span class="st">"qc_metrics_{sample_id}_before_filtering.png"</span>), <span class="at">plot =</span> qc_before, <span class="at">width =</span> <span class="dv">10</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb44-56"><a href="#cb44-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-57"><a href="#cb44-57" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot QC with image included</span></span>
<span id="cb44-58"><a href="#cb44-58" aria-hidden="true" tabindex="-1"></a>    p1 <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">annotate =</span> <span class="st">"sum"</span>, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">point_size =</span> <span class="fl">0.75</span>) <span class="sc">+</span> </span>
<span id="cb44-59"><a href="#cb44-59" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-60"><a href="#cb44-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-61"><a href="#cb44-61" aria-hidden="true" tabindex="-1"></a>    p2 <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">annotate =</span> <span class="st">"detected"</span>, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">point_size =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb44-62"><a href="#cb44-62" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-63"><a href="#cb44-63" aria-hidden="true" tabindex="-1"></a>    p3 <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">annotate =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">point_size =</span> <span class="fl">0.75</span>) <span class="sc">+</span> </span>
<span id="cb44-64"><a href="#cb44-64" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-65"><a href="#cb44-65" aria-hidden="true" tabindex="-1"></a>    tissue_plot <span class="ot">&lt;-</span> <span class="fu">plotVisium</span>(spe, <span class="at">highlight =</span> <span class="st">"in_tissue"</span>, <span class="at">spots =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb44-66"><a href="#cb44-66" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-67"><a href="#cb44-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-68"><a href="#cb44-68" aria-hidden="true" tabindex="-1"></a>    <span class="co"># display panels using patchwork</span></span>
<span id="cb44-69"><a href="#cb44-69" aria-hidden="true" tabindex="-1"></a>    qc_tissue_plot <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(p1, p2, p3, tissue_plot, <span class="at">nrow =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb44-70"><a href="#cb44-70" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"QC metrics for {sample_id} before filtering"</span>))</span>
<span id="cb44-71"><a href="#cb44-71" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(qc_tissue_plot)</span>
<span id="cb44-72"><a href="#cb44-72" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggsave</span>(<span class="fu">glue</span>(<span class="st">"qc_metrics_{sample_id}_tissue.png"</span>), qc_tissue_plot, <span class="at">width =</span> <span class="dv">20</span>, <span class="at">height =</span> <span class="dv">5</span>, <span class="at">dpi =</span> <span class="dv">200</span>)</span>
<span id="cb44-73"><a href="#cb44-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-74"><a href="#cb44-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-75"><a href="#cb44-75" aria-hidden="true" tabindex="-1"></a>    <span class="co"># determine outliers via thresholding on MAD from the median</span></span>
<span id="cb44-76"><a href="#cb44-76" aria-hidden="true" tabindex="-1"></a>    ol <span class="ot">&lt;-</span> <span class="fu">perCellQCFilters</span>(spe, <span class="at">sub.fields=</span><span class="st">"subsets_mito_percent"</span>)</span>
<span id="cb44-77"><a href="#cb44-77" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add results as cell metadata</span></span>
<span id="cb44-78"><a href="#cb44-78" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colData</span>(spe)[<span class="fu">names</span>(ol)] <span class="ot">&lt;-</span> ol </span>
<span id="cb44-79"><a href="#cb44-79" aria-hidden="true" tabindex="-1"></a>    <span class="co"># tabulate # and % of cells that'd </span></span>
<span id="cb44-80"><a href="#cb44-80" aria-hidden="true" tabindex="-1"></a>    <span class="co"># be discarded for different reasons</span></span>
<span id="cb44-81"><a href="#cb44-81" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb44-82"><a href="#cb44-82" aria-hidden="true" tabindex="-1"></a>        <span class="at">check.names=</span><span class="cn">FALSE</span>,</span>
<span id="cb44-83"><a href="#cb44-83" aria-hidden="true" tabindex="-1"></a>        <span class="st">`</span><span class="at">#</span><span class="st">`</span><span class="ot">=</span><span class="fu">apply</span>(ol, <span class="dv">2</span>, sum), </span>
<span id="cb44-84"><a href="#cb44-84" aria-hidden="true" tabindex="-1"></a>        <span class="st">`</span><span class="at">%</span><span class="st">`</span><span class="ot">=</span><span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span><span class="fu">apply</span>(ol, <span class="dv">2</span>, mean), <span class="dv">2</span>))</span>
<span id="cb44-85"><a href="#cb44-85" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot what spots will be discarded with MAD</span></span>
<span id="cb44-86"><a href="#cb44-86" aria-hidden="true" tabindex="-1"></a>    outlier_plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(ol), \(x) <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> x) <span class="sc">+</span> <span class="fu">ggtitle</span>(x))</span>
<span id="cb44-87"><a href="#cb44-87" aria-hidden="true" tabindex="-1"></a>    outlier_wrap <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(outlier_plots, <span class="at">nrow=</span><span class="dv">1</span>, <span class="at">guides=</span><span class="st">"collect"</span>) <span class="sc">+</span></span>
<span id="cb44-88"><a href="#cb44-88" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"Spots flagged for discard {sample_id}"</span>)) <span class="sc">&amp;</span></span>
<span id="cb44-89"><a href="#cb44-89" aria-hidden="true" tabindex="-1"></a>        <span class="fu">guides</span>(<span class="at">col=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">3</span>))) <span class="sc">&amp;</span></span>
<span id="cb44-90"><a href="#cb44-90" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_color_manual</span>(<span class="st">"discard"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="st">"lavender"</span>, <span class="st">"purple"</span>)) <span class="sc">&amp;</span></span>
<span id="cb44-91"><a href="#cb44-91" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="dv">8</span>), <span class="at">legend.key.size=</span><span class="fu">unit</span>(<span class="fl">0.8</span>, <span class="st">"lines"</span>))</span>
<span id="cb44-92"><a href="#cb44-92" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(outlier_wrap)</span>
<span id="cb44-93"><a href="#cb44-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-94"><a href="#cb44-94" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Set the threshold manually </span></span>
<span id="cb44-95"><a href="#cb44-95" aria-hidden="true" tabindex="-1"></a>    <span class="co"># select QC thresholds</span></span>
<span id="cb44-96"><a href="#cb44-96" aria-hidden="true" tabindex="-1"></a>    qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">600</span></span>
<span id="cb44-97"><a href="#cb44-97" aria-hidden="true" tabindex="-1"></a>    qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">400</span></span>
<span id="cb44-98"><a href="#cb44-98" aria-hidden="true" tabindex="-1"></a>    <span class="co"># qc_mito &lt;- colData(spe)$subsets_mito_percent &gt; 30</span></span>
<span id="cb44-99"><a href="#cb44-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-100"><a href="#cb44-100" aria-hidden="true" tabindex="-1"></a>    <span class="co"># number of discarded spots for each metric</span></span>
<span id="cb44-101"><a href="#cb44-101" aria-hidden="true" tabindex="-1"></a>    <span class="fu">apply</span>(<span class="fu">cbind</span>(qc_lib_size, qc_detected), <span class="dv">2</span>, sum)</span>
<span id="cb44-102"><a href="#cb44-102" aria-hidden="true" tabindex="-1"></a>    <span class="co"># combined set of discarded spots</span></span>
<span id="cb44-103"><a href="#cb44-103" aria-hidden="true" tabindex="-1"></a>    discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected</span>
<span id="cb44-104"><a href="#cb44-104" aria-hidden="true" tabindex="-1"></a>    <span class="fu">table</span>(discard)</span>
<span id="cb44-105"><a href="#cb44-105" aria-hidden="true" tabindex="-1"></a>    <span class="co"># discard spots</span></span>
<span id="cb44-106"><a href="#cb44-106" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colData</span>(spe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span>
<span id="cb44-107"><a href="#cb44-107" aria-hidden="true" tabindex="-1"></a>    <span class="co"># filter low-quality spots</span></span>
<span id="cb44-108"><a href="#cb44-108" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> spe[, <span class="sc">!</span><span class="fu">colData</span>(spe)<span class="sc">$</span>discard]</span>
<span id="cb44-109"><a href="#cb44-109" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dim</span>(spe)</span>
<span id="cb44-110"><a href="#cb44-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-111"><a href="#cb44-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-112"><a href="#cb44-112" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot the same QC on the spatial coordinates</span></span>
<span id="cb44-113"><a href="#cb44-113" aria-hidden="true" tabindex="-1"></a>    p1 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"sum"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span>  </span>
<span id="cb44-114"><a href="#cb44-114" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-115"><a href="#cb44-115" aria-hidden="true" tabindex="-1"></a>    p2 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"detected"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb44-116"><a href="#cb44-116" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-117"><a href="#cb44-117" aria-hidden="true" tabindex="-1"></a>    p3 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">pal =</span> <span class="st">"viridis"</span>, <span class="at">point_size =</span> <span class="fl">0.03</span>) <span class="sc">+</span> </span>
<span id="cb44-118"><a href="#cb44-118" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_void</span>()</span>
<span id="cb44-119"><a href="#cb44-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-120"><a href="#cb44-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-121"><a href="#cb44-121" aria-hidden="true" tabindex="-1"></a>    qc_after <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(p1, p2, p3, <span class="at">nrow =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb44-122"><a href="#cb44-122" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"QC metrics for {sample_id} after filtering"</span>))</span>
<span id="cb44-123"><a href="#cb44-123" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(qc_after)</span>
<span id="cb44-124"><a href="#cb44-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">glue</span>(<span class="st">"qc_metrics_{sample_id}_after_filtering.png"</span>), <span class="at">plot =</span> qc_after, <span class="at">width =</span> <span class="dv">10</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb44-125"><a href="#cb44-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-126"><a href="#cb44-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-127"><a href="#cb44-127" aria-hidden="true" tabindex="-1"></a>    <span class="do">### Processing </span></span>
<span id="cb44-128"><a href="#cb44-128" aria-hidden="true" tabindex="-1"></a>    <span class="co"># calculate logcounts and store in object</span></span>
<span id="cb44-129"><a href="#cb44-129" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb44-130"><a href="#cb44-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-131"><a href="#cb44-131" aria-hidden="true" tabindex="-1"></a>    <span class="co"># remove mitochondrial genes</span></span>
<span id="cb44-132"><a href="#cb44-132" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> spe[<span class="sc">!</span>is_mito, ]</span>
<span id="cb44-133"><a href="#cb44-133" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dim</span>(spe)</span>
<span id="cb44-134"><a href="#cb44-134" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb44-135"><a href="#cb44-135" aria-hidden="true" tabindex="-1"></a>    <span class="co"># fit mean-variance relationship</span></span>
<span id="cb44-136"><a href="#cb44-136" aria-hidden="true" tabindex="-1"></a>    dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(spe)</span>
<span id="cb44-137"><a href="#cb44-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-138"><a href="#cb44-138" aria-hidden="true" tabindex="-1"></a>    <span class="co"># select top HVGs</span></span>
<span id="cb44-139"><a href="#cb44-139" aria-hidden="true" tabindex="-1"></a>    top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb44-140"><a href="#cb44-140" aria-hidden="true" tabindex="-1"></a>    <span class="fu">length</span>(top_hvgs)</span>
<span id="cb44-141"><a href="#cb44-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-142"><a href="#cb44-142" aria-hidden="true" tabindex="-1"></a>    <span class="co"># run PCA and UMAP</span></span>
<span id="cb44-143"><a href="#cb44-143" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(spe, <span class="at">subset_row=</span>top_hvgs)</span>
<span id="cb44-144"><a href="#cb44-144" aria-hidden="true" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred=</span><span class="st">"PCA"</span>, <span class="at">pca =</span> <span class="dv">30</span>) <span class="co"># use 30 PCs as in the paper</span></span>
<span id="cb44-145"><a href="#cb44-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-146"><a href="#cb44-146" aria-hidden="true" tabindex="-1"></a>    <span class="co"># build shared nearest-neighbor (SNN) graph</span></span>
<span id="cb44-147"><a href="#cb44-147" aria-hidden="true" tabindex="-1"></a>    g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(spe, <span class="at">use.dimred=</span><span class="st">"PCA"</span>, <span class="at">type=</span><span class="st">"jaccard"</span>)</span>
<span id="cb44-148"><a href="#cb44-148" aria-hidden="true" tabindex="-1"></a>    <span class="co"># cluster via Leiden community detection algorithm</span></span>
<span id="cb44-149"><a href="#cb44-149" aria-hidden="true" tabindex="-1"></a>    k <span class="ot">&lt;-</span> <span class="fu">cluster_leiden</span>(g, <span class="at">objective_function=</span><span class="st">"modularity"</span>, <span class="at">resolution=</span><span class="fl">0.8</span>)</span>
<span id="cb44-150"><a href="#cb44-150" aria-hidden="true" tabindex="-1"></a>    <span class="fu">table</span>(spe<span class="sc">$</span>Leiden <span class="ot">&lt;-</span> <span class="fu">factor</span>(k<span class="sc">$</span>membership))</span>
<span id="cb44-151"><a href="#cb44-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-152"><a href="#cb44-152" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot clusters in spatial x-y coordinates</span></span>
<span id="cb44-153"><a href="#cb44-153" aria-hidden="true" tabindex="-1"></a>    spatial_p1 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">"Leiden"</span>, <span class="at">legend_position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb44-154"><a href="#cb44-154" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">unname</span>(pals<span class="sc">::</span><span class="fu">trubetskoy</span>()))  <span class="co"># Custom color palette for spatial plot</span></span>
<span id="cb44-155"><a href="#cb44-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-156"><a href="#cb44-156" aria-hidden="true" tabindex="-1"></a>    umap_p2 <span class="ot">&lt;-</span> <span class="fu">plotDimRed</span>(spe, <span class="at">plot_type =</span> <span class="st">"UMAP"</span>, </span>
<span id="cb44-157"><a href="#cb44-157" aria-hidden="true" tabindex="-1"></a>            <span class="at">annotate =</span> <span class="st">"Leiden"</span>) <span class="sc">+</span></span>
<span id="cb44-158"><a href="#cb44-158" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">unname</span>(pals<span class="sc">::</span><span class="fu">trubetskoy</span>()))  <span class="co"># Custom color palette for spatial plot</span></span>
<span id="cb44-159"><a href="#cb44-159" aria-hidden="true" tabindex="-1"></a>    cluster_plot <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(spatial_p1, umap_p2, <span class="at">nrow=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb44-160"><a href="#cb44-160" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="fu">glue</span>(<span class="st">"Leiden clusters for {sample_id}"</span>))</span>
<span id="cb44-161"><a href="#cb44-161" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(cluster_plot)</span>
<span id="cb44-162"><a href="#cb44-162" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">glue</span>(<span class="st">"clusters_{sample_id}.png"</span>), <span class="at">plot =</span> cluster_plot, <span class="at">width =</span> <span class="dv">10</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb44-163"><a href="#cb44-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-164"><a href="#cb44-164" aria-hidden="true" tabindex="-1"></a>    <span class="co"># save the object</span></span>
<span id="cb44-165"><a href="#cb44-165" aria-hidden="true" tabindex="-1"></a>    <span class="fu">saveRDS</span>(spe, <span class="at">file =</span> <span class="fu">glue</span>(<span class="st">"../data/ST_{sample_id}.rds"</span>))</span>
<span id="cb44-166"><a href="#cb44-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-167"><a href="#cb44-167" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span> (spe)</span>
<span id="cb44-168"><a href="#cb44-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-169"><a href="#cb44-169" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-63" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plobably plot QC of histograms separately and then use this function to plot each plot for each sample</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>spe_1 <span class="ot">&lt;-</span> <span class="fu">preprocess_sample</span>(sample_dirs[<span class="dv">1</span>])</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">basename</span>(sample_dirs[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-46-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="01_spatial_analysis.R_files/figure-html/cell-46-output-2.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-46-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="01_spatial_analysis.R_files/figure-html/cell-46-output-3.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-46-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="01_spatial_analysis.R_files/figure-html/cell-46-output-4.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-46-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="01_spatial_analysis.R_files/figure-html/cell-46-output-6.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<strong>pdf:</strong> 2
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-46-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="01_spatial_analysis.R_files/figure-html/cell-46-output-9.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-64" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot QC on the spatial coordinates</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>))</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_1)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_1)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_1)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">"QC histograms for {sample_id}"</span>), <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-47-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="01_spatial_analysis.R_files/figure-html/cell-47-output-1.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-65" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="14">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># process 2 sample</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>spe_2 <span class="ot">&lt;-</span> <span class="fu">preprocess_sample</span>(sample_dirs[<span class="dv">2</span>])</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">basename</span>(sample_dirs[<span class="dv">2</span>])</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="co"># histograms of QC metrics</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>))  <span class="co"># layout</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_2)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_2)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_2)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">"QC histograms for {sample_id}"</span>), <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-48-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="01_spatial_analysis.R_files/figure-html/cell-48-output-2.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-48-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="01_spatial_analysis.R_files/figure-html/cell-48-output-3.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-48-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="01_spatial_analysis.R_files/figure-html/cell-48-output-4.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-48-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="01_spatial_analysis.R_files/figure-html/cell-48-output-6.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-48-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="01_spatial_analysis.R_files/figure-html/cell-48-output-8.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-48-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="01_spatial_analysis.R_files/figure-html/cell-48-output-9.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-66" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="15">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># process 3 sample</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>spe_3 <span class="ot">&lt;-</span> <span class="fu">preprocess_sample</span>(sample_dirs[<span class="dv">3</span>])</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">basename</span>(sample_dirs[<span class="dv">3</span>])</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="co"># histograms of QC metrics</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>))  <span class="co"># layout</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_3)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_3)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_3)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">"QC histograms for {sample_id}"</span>), <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-49-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="01_spatial_analysis.R_files/figure-html/cell-49-output-2.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-49-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="01_spatial_analysis.R_files/figure-html/cell-49-output-3.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-49-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="01_spatial_analysis.R_files/figure-html/cell-49-output-4.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-49-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="01_spatial_analysis.R_files/figure-html/cell-49-output-6.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-49-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="01_spatial_analysis.R_files/figure-html/cell-49-output-8.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-49-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="01_spatial_analysis.R_files/figure-html/cell-49-output-9.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-67" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="16">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># process 4th sample</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>spe_4 <span class="ot">&lt;-</span> <span class="fu">preprocess_sample</span>(sample_dirs[<span class="dv">4</span>])</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">basename</span>(sample_dirs[<span class="dv">4</span>])</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="co"># histograms of QC metrics</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>))  <span class="co"># layout</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_4)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_4)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_4)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">"QC histograms for {sample_id}"</span>), <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-50-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="01_spatial_analysis.R_files/figure-html/cell-50-output-2.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-50-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="01_spatial_analysis.R_files/figure-html/cell-50-output-3.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-50-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="01_spatial_analysis.R_files/figure-html/cell-50-output-4.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-50-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="01_spatial_analysis.R_files/figure-html/cell-50-output-6.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-50-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="01_spatial_analysis.R_files/figure-html/cell-50-output-8.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-50-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-24"><img src="01_spatial_analysis.R_files/figure-html/cell-50-output-9.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-68" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="17">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># process 5th sample</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>spe_5 <span class="ot">&lt;-</span> <span class="fu">preprocess_sample</span>(sample_dirs[<span class="dv">5</span>])</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">basename</span>(sample_dirs[<span class="dv">5</span>])</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="co"># histograms of QC metrics </span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>))  <span class="co"># layout</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_5)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_5)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_5)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">"QC histograms for {sample_id}"</span>), <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-51-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-25"><img src="01_spatial_analysis.R_files/figure-html/cell-51-output-2.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-51-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26"><img src="01_spatial_analysis.R_files/figure-html/cell-51-output-3.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-51-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-27"><img src="01_spatial_analysis.R_files/figure-html/cell-51-output-4.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-51-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-28"><img src="01_spatial_analysis.R_files/figure-html/cell-51-output-6.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-51-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-29"><img src="01_spatial_analysis.R_files/figure-html/cell-51-output-8.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-51-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-30"><img src="01_spatial_analysis.R_files/figure-html/cell-51-output-9.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-69" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;vscode&quot;,&quot;value&quot;:{&quot;languageId&quot;:&quot;r&quot;}}" data-execution_count="19">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># process 6th sample</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>spe_6 <span class="ot">&lt;-</span> <span class="fu">preprocess_sample</span>(sample_dirs[<span class="dv">6</span>])</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">basename</span>(sample_dirs[<span class="dv">6</span>])</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a><span class="co"># histograms of QC metrics</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>), <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>))  <span class="co"># layout</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_6)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">"sum"</span>, <span class="at">main =</span> <span class="st">"UMIs per spot"</span>)</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_6)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">"detected"</span>, <span class="at">main =</span> <span class="st">"Genes per spot"</span>)</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe_6)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">"percent mitochondrial"</span>, <span class="at">main =</span> <span class="st">"Percent mito UMIs"</span>)</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">"QC histograms for {sample_id}"</span>), <span class="at">outer =</span> <span class="cn">TRUE</span>, <span class="at">cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-52-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-31"><img src="01_spatial_analysis.R_files/figure-html/cell-52-output-2.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-52-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-32"><img src="01_spatial_analysis.R_files/figure-html/cell-52-output-3.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-52-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-33"><img src="01_spatial_analysis.R_files/figure-html/cell-52-output-4.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.

Scale for <span class="ansi-green-fg">colour</span> is already present.

Adding another scale for <span class="ansi-green-fg">colour</span>, which will replace the existing scale.
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-52-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-34"><img src="01_spatial_analysis.R_files/figure-html/cell-52-output-6.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Saving 10 x 7 in image
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-52-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-35"><img src="01_spatial_analysis.R_files/figure-html/cell-52-output-8.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="01_spatial_analysis.R_files/figure-html/cell-52-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-36"><img src="01_spatial_analysis.R_files/figure-html/cell-52-output-9.png" width="420" height="420" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/sviatkh\.github\.io\/spatial-transcriptomics-project\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Spatial transcriptomics analysis</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>